### PR TITLE
LRU cache for octree nodes

### PR DESCRIPTION
**Pull request by thomasschiwietz**

### CODE REVIEW

File: sdl_viewer/src/main.rs [hunk 1]
```diff
@@ -135,6 +136,7 @@ struct NodeView {
     vertex_array: GlVertexArray,
     _buffer_position: GlBuffer,
     _buffer_color: GlBuffer,
+    used_memory: usize,
```
Comment (SirVer) [inline sdl_viewer/src/main.rs]: add unit? `used_memory_bytes`

File: sdl_viewer/src/main.rs [hunk 2]
```diff
@@ -215,13 +217,14 @@ impl NodeView {
...
             meta: node_data.meta,
+            used_memory: position.len() + color.len(),
         }
     }
 }
 
 // Keeps track of the nodes that were requested in-order and loads then one by one on request.
 struct NodeViewContainer {
-    node_views: HashMap<octree::NodeId, NodeView>,
+    cache: LruCache<octree::NodeId, NodeView>,
```
Comment (SirVer) [inline sdl_viewer/src/main.rs]: why did you rename this to cache? It still is conceptually the `node_views` hashmap

File: sdl_viewer/src/main.rs [hunk 3]
```diff
@@ -230,7 +233,7 @@ struct NodeViewContainer {
 }
 
 impl NodeViewContainer {
-    fn new(octree: Arc<octree::Octree>) -> Self {
+    fn new(octree: Arc<octree::Octree>, cache_size: usize) -> Self {
```
Comment (SirVer) [inline sdl_viewer/src/main.rs]: rename: `max_nodes_in_memory` to make clear what the unit of the cache_size is

File: sdl_viewer/src/main.rs [hunk 4]
```diff
@@ -246,8 +249,9 @@ impl NodeViewContainer {
                 node_data_sender.send((node_id, node_data)).unwrap();
             } 
         });
+        println!("Cache size {}", cache_size);
```
Comment (SirVer) [inline sdl_viewer/src/main.rs]: remove or make more descriptive, including units.

File: sdl_viewer/src/main.rs [hunk 5]
```diff
@@ -260,38 +264,40 @@ impl NodeViewContainer {
...
-        match self.node_views.entry(*node_id) {
-            Entry::Vacant(_) => {
-                // Limit the number of requested nodes because after a camera move
-                // requested nodes might not be in the frustum anymore.
-                if !self.requested.contains(&node_id) && self.requested.len() < 10 {  
-                    self.requested.insert(*node_id);
-                    self.node_id_sender.send(*node_id).unwrap();
-                }
-                None
+        if !self.cache.contains_key(node_id) {
```
Comment (SirVer) [inline sdl_viewer/src/main.rs]: Restructure to avoid right drift: if `self.cache.contains_key() { return ... };` then do the other case without else clause.

File: sdl_viewer/src/main.rs [hunk 6]
```diff
@@ -260,38 +264,40 @@ impl NodeViewContainer {
...
+                if !self.requested.contains(&node_id) {
+                    self.requested.insert(node_id);
+                    self.node_id_sender.send(node_id).unwrap();
                 }
-                Entry::Occupied(_) => {},
             }
         }
     }
+
+    fn get_used_memory(&self) -> usize {
```
Comment (SirVer) [inline sdl_viewer/src/main.rs]: get_used_memory_in_bytes?

File: sdl_viewer/src/main.rs [hunk 7]
```diff
@@ -302,10 +308,22 @@ fn main() {
                     .help("Input directory of the octree directory to serve.")
                     .index(1)
                     .required(true),
+                clap::Arg::with_name("cache_size")
```
Comment (SirVer) [inline sdl_viewer/src/main.rs]: Let's take MB here and use our heuristic to convert to nodes.

File: sdl_viewer/src/main.rs [hunk 8]
```diff
@@ -302,10 +308,22 @@ fn main() {
...
+                    .help("Maximum number of octree nodes to store in GPU memory")
+                    .required(false)
             ]
         )
         .get_matches();
 
+    // cache size
+    let mut cache_size = 10000;
```
Comment (SirVer) [inline sdl_viewer/src/main.rs]: `let mut cache_size = min(5000, matches.value_of("cache_size").unwrap_or("10000").map(|s| s.parse().unwrap()))` should work.

File: sdl_viewer/src/main.rs [hunk 9]
```diff
@@ -421,6 +439,9 @@ fn main() {
                 camera.height,
                 octree::UseLod::Yes,
             );
+            if visible_nodes.len() > cache_size {
+                visible_nodes.truncate(cache_size);
```
Comment (SirVer) [inline sdl_viewer/src/main.rs]: only truncate, no need for the if-statement. See docs https://doc.rust-lang.org/std/vec/struct.Vec.html#method.truncate

File: sdl_viewer/src/main.rs [hunk 10]
```diff
@@ -260,38 +264,40 @@ impl NodeViewContainer {
...
-                // requested nodes might not be in the frustum anymore.
-                if !self.requested.contains(&node_id) && self.requested.len() < 10 {  
-                    self.requested.insert(*node_id);
-                    self.node_id_sender.send(*node_id).unwrap();
-                }
-                None
-            }
-            Entry::Occupied(e) => Some(e.into_mut()),
+        if self.node_views.contains_key(node_id) {
+            return { self.node_views.get_mut(node_id).map(|f| f as &NodeView) }
```
Comment (SirVer) [inline sdl_viewer/src/main.rs]: remove {}?

File: sdl_viewer/src/main.rs [hunk 11]
```diff
@@ -260,38 +264,40 @@ impl NodeViewContainer {
...
+                    self.node_id_sender.send(node_id).unwrap();
                 }
-                Entry::Occupied(_) => {},
             }
         }
     }
+
+    fn get_used_memory_bytes(&self) -> usize {
+        let mut used_memory_bytes = 0;
+        for (_node_id, node_view) in self.node_views.iter() {
```
Comment (SirVer) [inline sdl_viewer/src/main.rs]: shorter and equivalent: `for (_, node_view) in &self.node_views {`

Comment (SirVer) [reply sdl_viewer/src/main.rs]: or more functional:

~~~
self.node_views().iter().map(|_, node_view| node_view.used_memory_bytes).sum()
~~~

File: sdl_viewer/src/main.rs [hunk 12]
```diff
@@ -302,10 +308,18 @@ fn main() {
...
                     .required(true),
+                clap::Arg::with_name("cache_size_mb")
+                    .help("Maximum cache size in MB for octree nodes in GPU memory")
+                    .required(false)
             ]
         )
         .get_matches();
 
+    // Maximum number of MB for the octree node cache in range 1..16 GB. The default is 2 GB
+    let max_mb_nodes: usize = cmp::min(cmp::max(1000, matches.value_of("cache_size_mb").unwrap_or("2000").parse().unwrap()), 16000);
```
Comment (SirVer) [inline sdl_viewer/src/main.rs]: this now is hard to read. suggestion:

~~~
let max_mb_nodes = {
     let value = matches.value_of("cache_size_mb").unwrap_or("2000").parse().unwrap();
     cmp::min(cmp::max(value, 1000), 16000))
};
~~~

**Comment by SirVer** [comment]
@wally-the-cartographer merge

**Comment by wally-the-cartographer** [comment]
Merge requested by authorized user SirVer. Merge queue now has a length of 2.

**Comment by wally-the-cartographer** [comment]
Error: exit status 1

Log:
~~~
=> handle work item: PR 60 on googlecartographer/point_cloud_viewer.
 => Switching to branch master.
 $ git reset --hard HEAD [data/point_cloud_viewer]
HEAD is now at 62dfbaa Build and test SDLViewer on Travis (#61)
 $ git clean -fd [data/point_cloud_viewer]
 $ git checkout master [data/point_cloud_viewer]
Already on 'master'
Your branch is up to date with 'origin/master'.
 $ git pull [data/point_cloud_viewer]
From https://github.com/googlecartographer/point_cloud_viewer
   62dfbaa..5725261  master     -> origin/master
Updating 62dfbaa..5725261
Fast-forward
 sdl_viewer/build.rs          |   4 +-
 sdl_viewer/src/box_drawer.rs |  62 ++++++++++-----------
 sdl_viewer/src/camera.rs     |  11 ++--
 sdl_viewer/src/glhelper.rs   |  46 +++++++--------
 sdl_viewer/src/graphic.rs    |  68 +++++++++++-----------
 sdl_viewer/src/lib.rs        |   2 +-
 sdl_viewer/src/main.rs       | 130 +++++++++++++++++++++----------------------
 web_viewer/Cargo.lock        |   8 +--
 8 files changed, 166 insertions(+), 165 deletions(-)
Current branch master is up to date.
 => Checking out thomasschiwietz/lru as local branch pr_60.
 $ git remote add thomasschiwietz git@github.com:thomasschiwietz/point_cloud_viewer.git [data/point_cloud_viewer]
fatal: remote thomasschiwietz already exists.
 $ git fetch thomasschiwietz [data/point_cloud_viewer]
 $ git branch --track pr_60 thomasschiwietz/lru [data/point_cloud_viewer]
Branch 'pr_60' set up to track remote branch 'lru' from 'thomasschiwietz' by rebasing.
 => Switching to branch pr_60.
 $ git reset --hard HEAD [data/point_cloud_viewer]
HEAD is now at 5725261 Switch to using StructGenerator from glbindings (#59)
 $ git clean -fd [data/point_cloud_viewer]
 $ git checkout pr_60 [data/point_cloud_viewer]
Switched to branch 'pr_60'
Your branch is up to date with 'thomasschiwietz/lru'.
 $ git pull --force [data/point_cloud_viewer]
Already up to date.
Current branch pr_60 is up to date.
 => merge master.
 $ git merge master [data/point_cloud_viewer]
Auto-merging sdl_viewer/src/main.rs
CONFLICT (content): Merge conflict in sdl_viewer/src/main.rs
Automatic merge failed; fix conflicts and then commit the result.
 => Abandoning branch pr_60.
 => Switching to branch master.
 $ git reset --hard HEAD [data/point_cloud_viewer]
HEAD is now at 10d381f implement reviewer suggestions
 $ git clean -fd [data/point_cloud_viewer]
 $ git checkout master [data/point_cloud_viewer]
Switched to branch 'master'
Your branch is up to date with 'origin/master'.
 $ git branch -D pr_60 [data/point_cloud_viewer]
Deleted branch pr_60 (was 10d381f).

~~~

**Comment by SirVer** [comment]
@wally-the-cartographer merge

**Comment by wally-the-cartographer** [comment]
Merge requested by authorized user SirVer. Merge queue now has a length of 1.
